---

# CSCI 2820, Spring 2022
# Homework 2
# Due: Friday January 28, 2022 at 11:59 PM

<br> 

### Your name:

<br> 

## Problem 1

A unit mass moves on a straight line (in one dimension). The position of the mass at time $t$ (in seconds) is denoted by $s(t)$, and its derivatives (the velocity and acceleration) by $s'(t)$ and $s''(t)$. The position as a function of time can be determined from Newton's second law 
$$ s''(t) = F(t) $$
where $F(t)$ is the force applied at time $t$, and the initial conditions $s(0),s'(0)$. We assume $F(t)$ is piecewise-constant, and is kept constant iin intervals of one second. The sequence of forces $F(t)$, for $0 \leq t < 10$, can be represented by a 10-vector $f$, with 
$$F(t) = f_k, \ \ \ k-1 \leq t < k $$

---
**Derive expressions for the final velocity $s'(10)$. Show that $s'(10)$ is an affine function of $x$, and give 10-vector $a$ and constants $b$ for which:
$$s'(10) = a^Tf+b$$**

---
This means that the mapping from the applied force sequence to the final velocity is affine. 

Hint: You can use that $$s'(t) = s'(0) + \int_0^t F(\tau) \ d\tau,$$


---
A note about your writeup: You may use the markdown cell below to latex your solution. If you do this, you  will recieve +1 extra credit points for this problem. You may also handwrite your solution and use the markdown cell below to include an image of your written work (Edit -> Insert Image). There is no extra credit point for including an image of handwritten work.

---

$$s''(t)=F(t)\\
F(t) = f_k, k-1\leq t < k\\
s'(t) = s'(0) + \int_0^t F(\tau) \ d\tau \\
s'(10) = s'(0) +\int_0^{10} F(\tau) \ d\tau$$ 
F(t) is a piecewise-constant by given and is kept constant in intervals of one second by given. So,
$$s'(10) = s'(0) + \int_0^{1} F(\tau) \ d\tau + \int_0^{2} F(\tau) \ d\tau + ... + \int_0^{10} F(\tau) \ d\tau$$$$$$
Since $F(t) = f_k, k-1\leq t < k$, 
$$s'(10) = s'(0) + \int_0^{1} f_1 \ d\tau + \int_0^{2} f_2 \ d\tau + ... + \int_0^{10} f_{10} \ d\tau$$$$$$
$$s'(10) = s'(0) + f_1 + f_2+ ... +f_{10}$$
Knowing that F(t) is constant in one second intervals we can use an arbitrary variable x.
$$s'(10) = s'(0) +f_1(x)+f_2(x)+...+f_{10}(x)$$
Above is the final expression for the final velocity $s'(10)$


Now to show that $s'(10)$ is an affine function of x we take the expression:
$$s'(10) = s'(0) +f_1(x)+f_2(x)+...+f_{10}(x)$$
$$s'(10) = b + a^Tf$$
Since this is a linear function with a constant added, it is an affine function.
$$a^T = [1,1,1,1,1,1,1,1,1,1]$$
$a^T$ is a vector of 1's as each interval takes place over one second each.
\begin{align}
f &= \begin{bmatrix}
f_1\\
f_2\\
\vdots\\
f_{10}
\end{bmatrix}
\end{align}
$$b=s'(0)$$
Above $a$ and $b$ are defined.

---
## Problem 2

**Part A:** Show that the following function is a linear function by showing that it has the property of superposition. Here, assume that $\vec{x}$ is an arbitrary $n-$vector with $n \geq 10$.

$$f(\vec{x}) = x_2+x_4+x_6+x_8+x_{10} $$

$$f(\alpha\vec{x}+\beta\vec{y}) = f(\alpha\vec{x}) +f(\beta\vec{y}) $$
$$\alpha f(\vec{x})=\alpha[x_2+x_4+x_6+x_8+x_{10}]$$
$$\beta f(\vec{y})=\beta[y_2+y_4+y_6+y_8+y_{10}]$$

$$f(\alpha\vec{x}) +f(\beta\vec{y}) = \alpha x_2 + \alpha x_4 + \alpha x_6 + \alpha x_8 + \alpha x_{10} + \beta y_2 +\beta y_4 +\beta y_6 +\beta y_8 +\beta y_{10}$$
$$[\alpha x_2 + \beta y_2] + [\alpha x_4 +\beta y_4] +[\alpha x_6 +\beta y_6] +[\alpha x_8 +\beta y_8] +[\alpha x_{10}+ \beta y_{10}]$$
$$=f(\alpha\vec{x}+\beta\vec{y})$$

**Part B:** Now let $\vec{x}$ be the 20-vector given below.

$$ \vec{x} = (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20)$$

Here, $x_1 = 1, x_2 = 2, ..., x_{20} = 20$. As you are coding, keeep in mind that Python is 0-indexed.

Use Python to compute $f(\vec{x})$ strictly as given by the definition above. i.e. You will be summing up 5 elements.



In [12]:
x = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]
fx= x[1] + x[3]+ x[5] +x[7]+x[9]
print (fx)

30


**Part C:** Now evaluate $f(\vec{x})$ again, but this time use the inner product representation of the function. Do not use NumPy for this part. Instead use a loop to compute the inner product. Define $\vec{x}$ and $\vec{a}$ as simple python lists. As a check, you should get the same answer here as you did in **Part B**.

In [14]:
a = [0,1,0,1,0,1,0,1,0,1, 0,0,0,0,0,0,0,0,0,0]
x = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]
fx = 0
for i in range(len(a)):
    fx += a[i] * x[i]
print (fx)

30


**Part D:** Now evaluate $f(\vec{x})$ again, but this time use Numpy and it's built-in dot product function. See the documentation here for syntax: https://numpy.org/doc/stable/reference/generated/numpy.dot.html

In [17]:
import numpy as np
a = [0,1,0,1,0,1,0,1,0,1, 0,0,0,0,0,0,0,0,0,0]
x = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]
print (np.dot(x,a,out=None))

30


## Problem 3

Now, we will use SciPy Stats to help us find the linear regression cooefficients for the data set `cars.csv`. Note that there is nothing for you to do in Parts A and B other than read, execute the cells, and interrpret the output. You'll be making interpretations of the output in Parts C and D.

**Part A:** Import cars.csv

In [8]:
import pandas as pd

dfCars = pd.read_csv("cars.csv")
dfCars.head()

Car_Name  Year  Selling_Price  Present_Price  Kms_Driven Fuel_Type  \
0     ritz  2014           3.35           5.59       27000    Petrol   
1      sx4  2013           4.75           9.54       43000    Diesel   
2     ciaz  2017           7.25           9.85        6900    Petrol   
3  wagon r  2011           2.85           4.15        5200    Petrol   
4    swift  2014           4.60           6.87       42450    Diesel   

  Seller_Type Transmission  Owner  
0      Dealer       Manual      0  
1      Dealer       Manual      0  
2      Dealer       Manual      0  
3      Dealer       Manual      0  
4      Dealer       Manual      0

**Part B:** Run the following code to import and use the statsmodels package to generate a model.

In [9]:

import statsmodels.api as sm 

# Collect the features in a 2D array 
X = dfCars[["Kms_Driven","Year"]]

# Add a constant to the array for the intecept 
X = sm.add_constant(X)

# Collect the response data in an array 
y = dfCars["Selling_Price"]

# Fit the ordinary least-squares (OLS) model 
model = sm.OLS(y, X).fit()

/opt/conda/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


In [39]:
model.params

const        -1223.847081
Kms_Driven       0.000028
Year             0.609591
dtype: float64

**Part C:** The output from `model.params` given in **Part B** gives us everything we need to create a model of the form: $$ \hat{y} = x^T \beta + v$$

Here, "Kms_Driven" and "Year" are the two entries in our $x-$ vector. And the constant term (const) is our offset. With that in mind, give the specific model that has been generated for us.

$\hat{y}= [0.000028 , 0.609591]^T 
\begin{align}
\begin{bmatrix}
\beta_1\\
\beta_2\\ 
\end{bmatrix}
\end{align}
- 1223.847081$ 


**Part D:** Answer the following questions:
- What is your interpretation of the $\beta_1$ coefficient? 
- What is your interpretation of the $\beta_2$ coefficient.

- What quantity is our model predicting?

The $\beta_1$ coefficient is the amount of kilometers driven on a given car being sold.
The $\beta_2$ coefficient is the difference between the current year and the year the car was manufactured.
The quantity that the model is predicting is the selling price of a car.

## Problem 4

In this problem, we will again utilize the `cars.csv` data.

**Part A:** First we will isolate the `Fuel_Type` column. Execute the following cell to copy the `Fuel_Type` column from the dfCars dataframe to a variable also called `Fuel_Type`.

In [10]:
Fuel_Type = dfCars["Fuel_Type"]

**Part B:** If we think of `Fuel_Type` as a vector. How many entries does this vector have?


In [14]:
typesOfFuel = []
for i in range (len(Fuel_Type)):
    matching = False
    for j in range (len(typesOfFuel)):
        if Fuel_Type[i] == typesOfFuel[j]:
            matching = True
    if matching == False:
        typesOfFuel.append(Fuel_Type[i])
        
print(len(typesOfFuel))

3


3

**Part C:** Use the idea of inner products to calculate the number of `Petrol` cars that there are for sale.

A few hints:
- It might be helpful to use a boolean variable to indicate whether a specific entry is `Petrol` or `Diesel`. I recommend making a vector of Boolean values (0s and 1s) that encodes this information.
- Then do an inner product computation to answer the desired question.

In [17]:
import numpy as np
petrol = []
for i in range (len(Fuel_Type)):
    if Fuel_Type[i] == "Petrol":

        petrol.append(1)
    else:
        petrol.append(0)
temp =[]
for j in range (len(Fuel_Type)):
    temp.append(1)
print (np.dot(temp,petrol,out=None))

239


239
